### PR TITLE
multiple custom actions validators

### PR DESCRIPTION
**Pull request by psacc**
### Description of changes
* add support for Resources async validation in addition to sync validation
* add support for async validators timeout, caching and retry
* UrlValidators use async execution with caching and retry
* Multiple custom actions resources uses async UrlValidators with timeout (and caching + retry)

### Tests
* manual local test: cluster creation with multiple custom actions
* added coverage for multiple custom actions in tests/integration-tests/tests/pcluster_api/test_api.py (testing async multithreaded validators in lambda)
* local integration tests execution:
```yaml
test-suites:
  create:
    test_create.py::test_cluster_creation_with_problematic_preinstall_script:
      dimensions:
        - regions: ["eu-west-1"]
          instances: {{ common.INSTANCES_DEFAULT_ARM }}
          schedulers: ["slurm"]
          oss: ["alinux2"]
  cfn-init:
    test_cfn_init.py::test_replace_compute_on_failure:
      dimensions:
        - regions: ["eu-west-1"]
          instances: {{ common.INSTANCES_DEFAULT_X86 }}
          oss: ["alinux2"]
          schedulers: ["slurm"]
    test_cfn_init.py::test_install_args_quotes:
      dimensions:
        - regions: ["eu-west-1"]
          instances: {{ common.INSTANCES_DEFAULT_X86 }}
          oss: ["alinux2"]
          schedulers: ["slurm"]
  pcluster_api:
    test_api_infrastructure.py::test_api_infrastructure_with_default_parameters:
      dimensions:
        - regions: ["eu-west-1"]
    test_api.py::test_cluster_slurm:
      dimensions:
        - regions: ["eu-west-1"]
          instances: {{ common.INSTANCES_DEFAULT_X86 }}
          oss: ["alinux2"]
          schedulers: ["slurm"]
    test_api.py::test_cluster_awsbatch:
      dimensions:
        - regions: ["eu-west-1"]
          instances: {{ common.INSTANCES_DEFAULT_X86 }}
          oss: ["alinux2"]
          schedulers: ["awsbatch"]
```

### References
* https://github.com/aws/aws-parallelcluster/pull/5104
* https://github.com/aws/aws-parallelcluster-cookbook/pull/1895
* https://github.com/aws/aws-parallelcluster-cookbook/pull/1990

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cli/src/pcluster/utils.py [hunk 1]
```diff
@@ -450,3 +453,88 @@ def batch_by_property_callback(items, property_callback: Callable[..., int], bat
...
+        """
+        if retry_on_exception is None:
+            retry_on_exception = (Exception,)
+
+        def decorator(func):
+            @functools.wraps(func)
+            async def wrapper(*args, **kwargs):
+                attempt = 0
+                result = None
+                while attempt < stop_max_attempt_number + 1:
```
Comment (eantonin) [inline cli/src/pcluster/utils.py]: nitpick, non-blocking: `<=` would be more readable to me

File: cli/src/pcluster/utils.py
Comment (eantonin) [inline cli/src/pcluster/utils.py:1]: question, non-blocking: We should probably unit test those utility functions don't we?

**Review by eantonin — DISMISSED** [review]

File: cli/src/pcluster/utils.py
Comment (psacc) [reply cli/src/pcluster/utils.py:1]: added for the cache wrapper at least

File: tests/integration-tests/tests/pcluster_api/test_api.py [hunk 2]
```diff
@@ -121,26 +125,134 @@ def _ec2_wait(region, instances, waiter_type):
     waiter.wait(InstanceIds=instances)
 
 
+@pytest.fixture(scope="session", name="resources_dir")
+def resources_dir_fixture():
+    return Path(pkg_resources.resource_filename(__name__, "/../../resources"))
+
+
+@pytest.fixture(scope="session", name="policies_template_path")
```
Comment (eantonin) [inline tests/integration-tests/tests/pcluster_api/test_api.py:133]: This is already in `conftest.py` you should be able to reuse it

File: tests/integration-tests/tests/pcluster_api/test_api.py [hunk 3]
```diff
@@ -121,26 +125,134 @@ def _ec2_wait(region, instances, waiter_type):
     waiter.wait(InstanceIds=instances)
 
 
+@pytest.fixture(scope="session", name="resources_dir")
+def resources_dir_fixture():
```
Comment (eantonin) [inline tests/integration-tests/tests/pcluster_api/test_api.py:129]: nitpick, non-blocking: This method and some of the below ones are copied from another class, we may put them under the generic `conftest.py` to DRY

File: tests/integration-tests/tests/pcluster_api/test_api.py [hunk 4]
```diff
@@ -180,6 +292,40 @@ def _test_cluster_workflow(region, api_client, create_cluster, request, pcluster
     _test_delete_cluster(region, cluster_operations_client, cluster_name)
 
 
+def _create_script_sequence(event_name, https_url, s3_url):
+    sequence = []
+
+    for i in range(10):
+        if i % 2 == 0:
+            cache_affinity = i % 2
+            url = f"{https_url}&cache_affinity={cache_affinity}"
```
Comment (eantonin) [inline tests/integration-tests/tests/pcluster_api/test_api.py:294]: nitpick,non-blocking: you can just hardcode 0 in here

Or, if you want to have just the half of the https url with cache, this does not work

File: tests/integration-tests/tests/pcluster_api/test_api.py [hunk 5]
```diff
@@ -121,26 +125,134 @@ def _ec2_wait(region, instances, waiter_type):
...
+    statement.append(
+        {
+            "Action": ["s3:GetObject"],
+            "Effect": "Allow",
+            "Resource": {"Fn::Sub": f"arn:${{AWS::Partition}}:s3:::{custom_actions_bucket_name}/*"},
+        }
+    )
+    statement.append(
+        {
+            "Action": ["events:PutRule", "events:DeleteRule", "events:PutTargets", "events:RemoveTargets"],
```
Comment (eantonin) [inline tests/integration-tests/tests/pcluster_api/test_api.py]: nitpick, non blocking, you should not need this?

**Review by eantonin — DISMISSED** [review]
LGTM, great work

File: tests/integration-tests/tests/pcluster_api/test_api.py (hunk 5 above)
Comment (psacc) [reply tests/integration-tests/tests/pcluster_api/test_api.py]: removed

**Review by eantonin — APPROVED** [review]

File: tests/integration-tests/tests/pcluster_api/test_api.py (hunk 4 above)
Comment (psacc) [reply tests/integration-tests/tests/pcluster_api/test_api.py:294]: good catch, introduced after adding s3 urls, fixed

**Pull request merged**